File: backend/app/formatters.py
import re


# Parses CMS numeric fields into floats or None.
def parse_number(value) -> float | None:
    if value is None or value == "":
        return None
    try:
        num = float(value) if not isinstance(value, (int, float)) else float(value)
    except (TypeError, ValueError):
        return None
    if num != num or num in (float("inf"), float("-inf")):
        return None
    return num


# Builds a single-line facility address from CMS parts.
def format_address(address: str = "", city: str = "", state: str = "", zip_code: str = "") -> str:
    line1 = (address or "").strip()
    city_state_zip = ", ".join(p for p in [city, state, zip_code] if p).strip()
    if line1 and city_state_zip:
        return f"{line1}, {city_state_zip}"
    return line1 or city_state_zip or "N/A"


# Strips non-digits and normalizes CCN to six digits.
def normalize_ccn(ccn: str) -> str:
    digits = "".join(c for c in ccn if c.isdigit())
    return digits.zfill(6)[-6:]


# Returns whether the CCN is exactly six digits.
def is_valid_ccn(ccn: str) -> bool:
    normalized = normalize_ccn(ccn)
    return len(normalized) == 6 and normalized.isdigit()


# Builds the Medicare Care Compare URL for a CCN.
def medicare_care_compare_url(ccn: str) -> str:
    return f"https://www.medicare.gov/care-compare/details/nursing-home/{normalize_ccn(ccn)}"


# Title-cases CMS provider names for display.
def to_display_name(name: str) -> str:
    lowered = name.lower()
    return re.sub(r"\b([a-z])", lambda m: m.group(1).upper(), lowered)

File: frontend/lib/formatters.ts
export function parseNumber(value: unknown): number | null {
  if (value == null || value === "") return null;
  const num = typeof value === "number" ? value : parseFloat(String(value));
  return Number.isFinite(num) ? num : null;
}

export function formatAddress(parts: {
  address?: string;
  city?: string;
  state?: string;
  zip?: string;
}): string {
  const line1 = parts.address?.trim() ?? "";
  const cityStateZip = [parts.city, parts.state, parts.zip]
    .filter(Boolean)
    .join(", ")
    .replace(/,\s*,/g, ",")
    .trim();

  if (line1 && cityStateZip) {
    return `${line1}, ${cityStateZip}`;
  }
  return line1 || cityStateZip || "N/A";
}

export function normalizeCcn(ccn: string): string {
  return ccn.replace(/\D/g, "").padStart(6, "0").slice(-6);
}

export function isValidCcn(ccn: string): boolean {
  return /^\d{6}$/.test(normalizeCcn(ccn));
}

export function medicareCareCompareUrl(ccn: string): string {
  return `https://www.medicare.gov/care-compare/details/nursing-home/${normalizeCcn(ccn)}`;
}

export function toDisplayName(name: string): string {
  return name
    .toLowerCase()
    .replace(/\b([a-z])/g, (_, letter: string) => letter.toUpperCase());
}
